src/bvals/cc/sparse_bvals_cc_in_one.cpp

```diff
@@ -0,0 +1,534 @@
+      // Separate reflux buffer if needed
+      if ((nb.snb.level - 1 == pmb->loc.level) &&
+          (std::abs(nb.ni.ox1) + std::abs(nb.ni.ox2) + std::abs(nb.ni.ox3) == 1))
+        pmesh->boundary_comm_reflux_map[r_key] = CommBuffer<buf_pool_t<Real>::owner_t>(
+            tag, receiver_rank, sender_rank, comm_reflux, get_resource_method,
+            use_sparse_buffers);
+    }
+  });
+
+  Kokkos::Profiling::popRegion();
+  return TaskStatus::complete;
+}
+
+// Build a vector of pointers to all of the sending or receiving communication buffers on
+// MeshData md. This cache is important for performance, since this elides a map look up
+// for the buffer every time the bvals code iterates over boundaries.
+//
+// The buffers in the cache do not necessarily need to be in the same order as the
+// sequential order of the ForEachBoundary iteration. Therefore, this also builds a vector
+// for indexing from the sequential boundary index defined by the iteration pattern of
+// ForEachBoundary to the index of the buffer corresponding to this boundary in the buffer
+// cache. This allows for reordering the calls to send and receive on the buffers, so that
+// MPI_Isends and MPI_Irecvs get posted in the same order (approximately, due to the
+// possibility of multiple MeshData per rank) on the sending and receiving ranks. In
+// simple tests, this did not have a big impact on performance but I think it is useful to
+// leave the machinery here since it doesn't seem to have a big overhead associated with
+// it (LFR).
+template <BoundaryType bound_type, class V1, class V2, class F>
+void BuildBufferCache(std::shared_ptr<MeshData<Real>> &md, V1 *pbuf_vec, V2 *pidx_vec,
+                      F KeyFunc) {
+  Mesh *pmesh = md->GetMeshPointer();
+
+  using key_t = std::tuple<int, int, std::string, int>;
+  std::vector<std::tuple<int, int, key_t>> key_order;
+
+  int boundary_idx = 0;
+  ForEachBoundary<bound_type>(
+      md, [&](sp_mb_t pmb, sp_mbd_t rc, nb_t &nb, const sp_cv_t v) {
+        auto key = KeyFunc(pmb, nb, v);
+        PARTHENON_DEBUG_REQUIRE(pmesh->boundary_comm_map.count(key) > 0,
+                                "Boundary communicator does not exist");
+        // Create a unique index by combining receiver gid (second element of the key
+        // tuple) and geometric element index (fourth element of the key tuple)
+        int recvr_idx = 27 * std::get<1>(key) + std::get<3>(key);
+        key_order.push_back({recvr_idx, boundary_idx, key});
+        ++boundary_idx;
+      });
+
+  // If desired, sort the keys and boundary indices by receiver_idx
+  // std::sort(key_order.begin(), key_order.end(),
+  //          [](auto a, auto b) { return std::get<0>(a) < std::get<0>(b); });
+
+  // Or, what the hell, you could put them in random order if you want, which
+  // frighteningly seems to run faster in some cases
+  std::random_device rd;
+  std::mt19937 g(rd());
+  std::shuffle(key_order.begin(), key_order.end(), g);
```

Comment:
yeah, I think there are a number of MPI knobs to turn. It would definitely be interesting to more systematically look at how these things affect performance in more realistic setups. I also think there might be some advantage of putting the `TryReceive` call in a background thread so that it can continually poll.